### PR TITLE
add example4 - dmdarray adding

### LINKED ISSUE #619: add a tutorial example - adding 2 mdarrays

### PR DESCRIPTION
**Pull request by haichangsi**

### CODE REVIEW

File: src/example4.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// SPDX-FileCopyrightText: Intel Corporation
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <dr/mhp.hpp>
+#include <fmt/core.h>
+
+namespace mhp = dr::mhp;
+using T = int;
+
```
Comment (mateuszpn) [inline src/example4.cpp:10]: Please add a descriptive comment, eg.
/* The example presents operation of ... The result is stored in ... */

File: src/example4.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+int main(int argc, char **argv) {
+
+  mhp::init(sycl::default_selector_v);
+  std::size_t xdim = 9, ydim = 5;
+
+  std::array<std::size_t, 2> extents2d = {xdim, ydim};
+
+  mhp::distributed_mdarray<T, 2> a(extents2d);
+  mhp::distributed_mdarray<T, 2> b(extents2d);
+  mhp::distributed_mdarray<T, 2> c(extents2d);
```
Comment (mateuszpn) [inline src/example4.cpp:21]: Please add a comment to lines 21-22, encouraging a tutorial user to change the initial content of arrays a & b.

**Review by mateuszpn — CHANGES_REQUESTED** [review]
When running in system with Intel GPU (devcloud in particular) results in an error:

```
uab8f7faf7d5cc8a0d0c8bf0d3553a43@idc-beta-batch-head-node:~/work/tutorial-haichangsi$ I_MPI_OFFLOAD=0 mpirun -n 4 ./build/src/example4
[1699012212.351400] [idc-beta-batch-head-node:1845409:0]        ib_iface.c:1017 UCX  ERROR ibv_create_cq(cqe=4096) failed: Cannot allocate memory : Please set max locked memory (ulimit -l) to 'unlimited' (current: 4096 kbytes)
Abort(1615247) on node 0 (rank 0 in comm 0): Fatal error in PMPI_Init: Other MPI error, error stack:
MPIR_Init_thread(176)........:
MPID_Init(1548)..............:
MPIDI_OFI_mpi_init_hook(1632):
create_vni_context(2208).....: OFI endpoint open failed (ofi_init.c:2208:create_vni_context:Input/output error)
```
(tested with locally installed IMPI 2021.11). Please verify in different system/config. Maybe setting of some I_MPI_* env vars will help.

**Comment by haichangsi** [comment]
> When running in system with Intel GPU (devcloud in particular) results in an error:
> 
> ```
> uab8f7faf7d5cc8a0d0c8bf0d3553a43@idc-beta-batch-head-node:~/work/tutorial-haichangsi$ I_MPI_OFFLOAD=0 mpirun -n 4 ./build/src/example4
> [1699012212.351400] [idc-beta-batch-head-node:1845409:0]        ib_iface.c:1017 UCX  ERROR ibv_create_cq(cqe=4096) failed: Cannot allocate memory : Please set max locked memory (ulimit -l) to 'unlimited' (current: 4096 kbytes)
> Abort(1615247) on node 0 (rank 0 in comm 0): Fatal error in PMPI_Init: Other MPI error, error stack:
> MPIR_Init_thread(176)........:
> MPID_Init(1548)..............:
> MPIDI_OFI_mpi_init_hook(1632):
> create_vni_context(2208).....: OFI endpoint open failed (ofi_init.c:2208:create_vni_context:Input/output error)
> ```
> 
> (tested with locally installed IMPI 2021.11). Please verify in different system/config. Maybe setting of some I_MPI_* env vars will help.

as we talked, I think it's because of your specific MPI configuration

File: src/example4.cpp (hunk 1 above)
Comment (haichangsi) [reply src/example4.cpp:10]: I thought about describing all new examples together in the README file

File: src/example4.cpp (hunk 2 above)
Comment (haichangsi) [reply src/example4.cpp:21]: Done

File: src/example4.cpp (hunk 1 above)
Comment (mateuszpn) [reply src/example4.cpp:10]: So a brief summary, like 
```/* add content of two 2-d arrays and display the results */```
The files have their names (not informing about the content, just exampleX), so as a reader of a tutorial, I would appreciate some additional textual sync between the description and the particular code.

**Review by mateuszpn — APPROVED** [review]

**Pull request merged**